### ISSUE TITLE
[FO4Edit] VMADs with unusually structured script properties cannot be parsed

### ISSUE THREAD

**Issue by EyeDeck** [TES5Edit/TES5Edit#544]
**What is the version of xEdit you are using?**

FO4Edit 3.2.62 EXPERIMENTAL (D0DAD251)

**Describe the bug**

I'll start by saying that this _probably_ isn't worth fixing since I'd be surprised if anyone other than me will ever find this themselves--so if someone wants to tell me to "just stop doing that" or whatever I understand--but I'll post it anyway.

So, [according to the Papyrus documentation](https://www.creationkit.com/fallout4/index.php?title=Struct_Reference#Defining_a_Struct), it is illegal for a struct to contain either arrays or other structs, and if you try to do that then the Bethesda Papyrus compiler will throw an error and prevent it. Interestingly, this error is actually a completely artificial restriction imposed by the compiler, and if you have [a compiler that ignores this restriction](https://github.com/Orvid/Caprica), both the CK and the game engine actually have no problem using structs that are set up this way.

xEdit, however, does not know how to parse VMAD records containing predefined script properties that are set up in this manner, so if presented with one it will just start producing noise for the rest of the VMAD record.

**Expected behavior**

Ideally xEdit ought to be able to display a nice tree with properly parsed script properties.

**Screenshots**

From a test mod I made to demonstrate, with an intentionally horribly convoluted script property:
![ck](https://user-images.githubusercontent.com/2722970/45261136-6ee2f980-b3c8-11e8-99d4-23224f821fed.png)
xEdit failing to parse it:
![xedit](https://user-images.githubusercontent.com/2722970/45261137-828e6000-b3c8-11e8-9fc2-ca74d22568d7.png)

**Additional context**

Here's a test mod with an example of one of these script properties (same as the screenshots above):
[SillyScript.zip](https://github.com/TES5Edit/TES5Edit/files/2363975/SillyScript.zip)

The script used in this test looks like this:
```Papyrus
Scriptname SillyScript extends Quest

NestedStruct Property testStructA Auto Const
NestedStruct Property testStructB Auto

Struct NestedStruct
	NestedStruct recursion
	NestedStruct[] moreRecursion
	int level
EndStruct

NestedStruct Function Test(int levels, NestedStruct input = None)
	if (input == None)
		input = new NestedStruct
	endif
	if (levels == 0)
		return input
	endif
	
	input.level = levels
	
	input.recursion = new NestedStruct
	Test(levels-1, input.recursion)
	return input
EndFunction

Function RunTest()
	testStructB = Test(20)
	Debug.Trace("\n\nPrinting testStructA (editor defined):\n" + testStructA)
	Debug.Trace("\n\nPrinting testStructB (Papyrus defined):\n" + testStructB)
EndFunction

Function Crash()
	{Causes an instant CTD}
	NestedStruct killpill = new NestedStruct
	killpill.recursion = killpill
	Debug.Trace(killpill)
EndFunction
```

To test it yourself, load the plugin and in the console enter: `cqf __TestQuest RunTest`

Then, in Papyrus.0.log, you should be able to find a stringified version of these script properties:
```
Printing testStructA (editor defined):
[recursion = [recursion = [recursion = [recursion = [recursion = [recursion = [recursion = None, moreRecursion = [], level = 7], moreRecursion = [], level = 6], moreRecursion = [[recursion = None, moreRecursion = [], level = 6]], level = 5], moreRecursion = [], level = 4], moreRecursion = [], level = 3], moreRecursion = [], level = 2], moreRecursion = [[recursion = None, moreRecursion = [], level = 2], [recursion = None, moreRecursion = [], level = 2], [recursion = [recursion = None, moreRecursion = [], level = 3], moreRecursion = [], level = 2]], level = 1]

Printing testStructB (Papyrus defined):
[recursion = [recursion = [recursion = [recursion = [recursion = [recursion = [recursion = [recursion = [recursion = [recursion = [recursion = [recursion = [recursion = [recursion = [recursion = [recursion = [recursion = [recursion = [recursion = [recursion = [recursion = None, moreRecursion = [], level = 0], moreRecursion = [], level = 1], moreRecursion = [], level = 2], moreRecursion = [], level = 3], moreRecursion = [], level = 4], moreRecursion = [], level = 5], moreRecursion = [], level = 6], moreRecursion = [], level = 7], moreRecursion = [], level = 8], moreRecursion = [], level = 9], moreRecursion = [], level = 10], moreRecursion = [], level = 11], moreRecursion = [], level = 12], moreRecursion = [], level = 13], moreRecursion = [], level = 14], moreRecursion = [], level = 15], moreRecursion = [], level = 16], moreRecursion = [], level = 17], moreRecursion = [], level = 18], moreRecursion = [], level = 19], moreRecursion = [], level = 20]
```

Here testStructA was what I manually defined in the editor, and testStructB was the output of the recursive Test() function.

Anyway, I wouldn't worry too much about this if this is a difficult thing to fix, it's just that I was working on a plugin wherein I wanted to renumber some FormIDs that were referenced by a predefined Papyrus property (array of structs each containing another array of different structs, with the records being referenced by the innermost structs), and I found that xEdit couldn't _quite_ handle what I was trying to do, which is a very rare occurrence.

**Comment by eFrysTon** [comment]
It doesn't surprise me that that currently doesn't work, if you look at the definition:
```pascal
  wbScriptPropertyStruct :=
    wbArrayS('Struct', wbStructSK([0], 'Member', [
      wbLenString('memberName', 2),
      wbInteger('Type', itU8, wbPropTypeEnum, cpNormal, False, nil, wbScriptPropertyTypeAfterSet),
      wbInteger('Flags', itU8, wbEnum([
        {0x00} '',
        {0x01} 'Edited',
        {0x02} '',
        {0x03} 'Removed'
      ])).SetDefaultEditValue('Edited'),
      wbUnion('Value', wbScriptPropertyStructMemberDecider, [
        {00} wbNull,
        {01} wbScriptPropertyObject,
        {02} wbLenString('String', 2),
        {03} wbInteger('Int32', itS32),
        {04} wbFloat('Float'),
        {05} wbInteger('Bool', itU8, wbBoolEnum),
        {11} wbArray('Array of Object', wbScriptPropertyObject, -1),
        {12} wbArray('Array of String', wbLenString('Element', 2), -1),
        {13} wbArray('Array of Int32', wbInteger('Element', itS32), -1),
        {14} wbArray('Array of Float', wbFloat('Element'), -1),
        {15} wbArray('Array of Bool', wbInteger('Element', itU8, wbBoolEnum), -1)
      ])
    ]), -1, cpNormal, False);
```
you can see that a wbScriptPropertyStruct can't currently contain a wbScriptPropertyStruct as member.

Which makes sense, because if I were to put wbScriptPropertyStruct as a possible value into that wbUnion, then it would still be nil at that time (as we are in the middle of the statement which creates the value that is going to get assigned to wbScriptPropertyStruct).

Basically, as they are right now, the record definitions can't possibly handle recursions.

But I have an idea how I can make it work (based on the fact that the wbUnion is only resolved at runtime to a specific contained definition) so I might be able to delay the lookup until that takes place, by which time wbScriptPropertyStruct would be defined.

I'll look into that.

**Comment by eFrysTon** [comment]
https://github.com/TES5Edit/TES5Edit/compare/e4bd07027d3b...b9d4ad75cf76

Should fix this issue.

**Comment by EyeDeck** [comment]
I am impressed. Very nice work.
![image](https://user-images.githubusercontent.com/2722970/45267383-50691680-b439-11e8-92c5-9270c5a5d260.png)
